qubes-rpc/qubes.Syslog

```diff
@@ -0,0 +1,164 @@
+           non-printable characters and truncates messages at 1500 characters.
+           By default non-ASCII characters are replaced with question marks.
+           To allow unicode, set QubesDB `/qrexec/qubes-syslog/allow-unicode` to
+           `True`. This infers security risks with unicode parsing bugs in log
+           viewers.
+
+Please note that the internal message structure is not sanitized. So any receiving
+application should perform its own sanitization, if the message is structured
+further.
+
+For example, a user could deploy the following rsyslog configuration in a sending
+VM at `/etc/rsyslog.d/qubes_syslog.conf` (requires `systemctl restart rsyslog`):
+```
+module(load="omprog")
+
+template(name="qubes_syslog" type="string"
+    string="<%PRI%>%TIMESTAMP:::date-rfc3339% %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n")
+
+ruleset(name="qubes") {
+  action(
+    type="omprog"
+    binary="/usr/sbin/runuser -u user -- /usr/bin/qrexec-client-vm work qubes.Syslog"
+    template="qubes_syslog"
+  )
+}
+
+#apply your filtering here (*.* just forwards all logs)
+*.* call qubes
+```
+
+Users can also e.g. pipe stored log files to a VM via
+`qrexec-client-vm vm qubes.Syslog < /path/to/some/logfile`.
+
+References:
+[RFC3164]: https://datatracker.ietf.org/doc/html/rfc3164
+[RFC5424]: https://datatracker.ietf.org/doc/html/rfc5424
+'''
+
+import io
+import sys
+import os
+import logging
+import syslog
+import re
+import unicodedata
+import qubesdb
+
+def get_logger():
+    log = logging.getLogger("qubes.Syslog")
+    log.setLevel(logging.INFO)
+    return log
+
+#globals
+LOG = get_logger()
+SEV2SYSLOG = {
+    0: syslog.LOG_EMERG,
+    1: syslog.LOG_ALERT,
+    2: syslog.LOG_CRIT,
+    3: syslog.LOG_ERR,
+    4: syslog.LOG_WARNING,
+    5: syslog.LOG_NOTICE,
+    6: syslog.LOG_INFO,
+    7: syslog.LOG_DEBUG,
+}
+
+def extract_priority(pri):
+    severity = pri % 8
+    return SEV2SYSLOG[severity]
+
+def filter_non_printable(string):
+    ret = []
+    for ch in string:
+        cat = unicodedata.category(ch)
+        if cat[0] == "C":
+            ret += "?"
+        else:
+            ret += ch
+    return ''.join(ret)
+
+def sanitize(untrusted, allow_unicode=False):
+    untr = untrusted.rstrip("\n").rstrip("\r")
+
+    #truncate
+    untr = (untr[:1500] + '..') if len(untr) > 1500 else untr
+
+    m = re.match(r"^<([0-9]{1,3})>.*$", untr)
+    if m:
+        prio = extract_priority(int(m.group(1)))
+    else:
+        prio = syslog.LOG_INFO
+
+    if not allow_unicode:
+        untr = untr.encode(encoding="ascii", errors="replace").decode("ascii")
+
+    return prio, filter_non_printable(untr)
+
+def handle_untrusted(vm, allow_unicode=False):
+    pid = os.getpid()
+    syslog.openlog(ident=f"qubes.Syslog|{vm}[{pid}]", facility=syslog.LOG_USER)
+
+    input_stream = io.TextIOWrapper(sys.stdin.buffer, encoding="utf-8")
+    for untrusted_line in input_stream:
```

Comment:
This allows a trivial DoS by sending a very very very long line.

Reply:
Yup. I didn't bother about DoS as there's many opportunities in Qubes to do something like that (e.g. use the CPU a lot, I/O a lot, send tons of useless logs making journald rotate out the relevant ones, ...). Since it's easy to fix, I fixed it though.